File: android/src/main/kotlin/com/huddlecommunity/native_video_player/handlers/VideoPlayerQualityHandler.kt
package com.huddlecommunity.native_video_player.handlers

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URL

/**
 * Handles HLS quality parsing and management
 * Equivalent to iOS VideoPlayerQualityHandler
 */
object VideoPlayerQualityHandler {
    private const val TAG = "VideoPlayerQuality"

    /**
     * Fetches and parses HLS qualities from an M3U8 playlist URL
     * @param url The M3U8 playlist URL
     * @return List of quality maps with "label" and "url" keys
     */
    suspend fun fetchHLSQualities(url: String): List<Map<String, String>> = withContext(Dispatchers.IO) {
        try {
            val connection = URL(url).openConnection()
            val playlist = connection.getInputStream().bufferedReader().use { it.readText() }

            val qualities = mutableListOf<Map<String, String>>()
            val lines = playlist.lines()
            var lastResolution = ""

            for (line in lines) {
                when {
                    line.contains("#EXT-X-STREAM-INF") -> {
                        // Extract resolution using regex
                        val resolutionMatch = Regex("RESOLUTION=(\\d+x\\d+)").find(line)
                        lastResolution = resolutionMatch?.groupValues?.get(1) ?: ""
                    }
                    line.endsWith(".m3u8") && lastResolution.isNotEmpty() -> {
                        qualities.add(
                            mapOf(
                                "label" to lastResolution,
                                "url" to line
                            )
                        )
                        lastResolution = ""
                    }
                }
            }

            Log.d(TAG, "Parsed ${qualities.size} quality variants from HLS playlist")
            qualities
        } catch (e: Exception) {
            Log.e(TAG, "Error fetching HLS qualities: ${e.message}", e)
            emptyList()
        }
    }
}

File: ios/Classes/Handlers/VideoPlayerQualityHandler.swift
import Foundation

extension VideoPlayerView {
    func fetchHLSQualities(from url: URL, completion: @escaping ([[String: String]]) -> Void) {
        URLSession.shared.dataTask(with: url) { data, _, error in
            guard let data = data, error == nil,
                  let playlist = String(data: data, encoding: .utf8)
            else {
                completion([])
                return
            }

            var qualities: [[String: String]] = []
            let lines = playlist.components(separatedBy: "\n")
            var lastResolution = ""

            for line in lines {
                if line.contains("#EXT-X-STREAM-INF") {
                    if let resMatch = line.range(of: "RESOLUTION=\\d+x\\d+", options: .regularExpression) {
                        lastResolution = String(line[resMatch]).replacingOccurrences(of: "RESOLUTION=", with: "")
                    }
                } else if line.hasSuffix(".m3u8") {
                    qualities.append(["label": lastResolution, "url": line])
                }
            }

            completion(qualities)
        }
        .resume()
    }
}